sway-lib-std/src/u256.sw

```diff
@@ -88,4 +106,263 @@ impl U256 {
+        } else if w == 1 {
+            let(shifted_3, carry_3) = lsh_with_carry(word_3, b);
+            w1 = (word_2 << b) + carry_3;
+            let(shifted_4, carry_4) = lsh_with_carry(word_4, b);
+            w2 = shifted_3 + carry_4;
+            w3 = shifted_4;
+        } else if w == 2 {
+            let(shifted_4, carry_4) = lsh_with_carry(word_4, b);
+            w1 = (word_3 << b) + carry_4;
+            w2 = shifted_4;
+        } else if w == 3 {
+            w1 = word_4 << b;
+        } else {
+            ();
+        };
+
+        ~U256::from(w1, w2, w3, w4)
+    }
+
+    fn rsh(self, shift_amount: u64) -> Self {
+        let(word_1, word_2, word_3, word_4) = self.decompose();
+        let mut w1 = 0;
+        let mut w2 = 0;
+        let mut w3 = 0;
+        let mut w4 = 0;
+
+        let w = shift_amount / 64; // num of whole words to shift in addition to b
+        let b = shift_amount % 64; // num of bits to shift within each word
+
+        if w == 0 {
+            let(shifted_3, carry_3) = rsh_with_carry(word_3, b);
+            w4 = (word_4 >> b) + carry_3;
+            let(shifted_2, carry_2) = rsh_with_carry(word_2, b);
+            w3 = shifted_3 + carry_2;
+            let(shifted_1, carry_1) = rsh_with_carry(word_1, b);
+            w2 = shifted_2 + carry_1;
+            w1 = shifted_1;
+        } else if w == 1 {
+            let(shifted_2, carry_2) = rsh_with_carry(word_2, b);
+            w4 = (word_3 >> b) + carry_2;
+            let(shifted_1, carry_1) = rsh_with_carry(word_1, b);
+            w3 = shifted_2 + carry_1;
+            w2 = shifted_1;
+        } else if w == 2 {
+            let(shifted_1, carry_1) = rsh_with_carry(word_1, b);
+            w4 = (word_2 >> b) + carry_1;
+            w3 = shifted_1;
+        } else if w == 3 {
+            w4 = word_1 >> b;
+        };
+
+        ~U256::from(w1, w2, w3, w4)
+    }
+}
+
+impl core::ops::Add for U256 {
+    /// Add a U256 to a U256. Panics on overflow.
+    fn add(self, other: Self) -> Self {
+        let(word_1, word_2, word_3, word_4) = self.decompose();
+        let(other_word_1, other_word_2, other_word_3, other_word_4) = other.decompose();
+
+        let mut overflow = 0;
+        let mut local_res = ~U128::from(0, word_4) + ~U128::from(0, other_word_4);
+        let result_d = local_res.lower;
+        overflow = local_res.upper;
+
+        local_res = ~U128::from(0, word_3) + ~U128::from(0, other_word_3) + ~U128::from(0, overflow);
+        let result_c = local_res.lower;
+        overflow = local_res.upper;
+
+        local_res = ~U128::from(0, word_2) + ~U128::from(0, other_word_2) + ~U128::from(0, overflow);
+        let result_b = local_res.lower;
+        overflow = local_res.upper;
+
+        local_res = ~U128::from(0, word_1) + ~U128::from(0, other_word_1) + ~U128::from(0, overflow);
+        let result_a = local_res.lower;
+        if local_res.upper != 0 {
+            // panic on overflow
+            revert(0);
+        };
```

Comment:
This could probably be an `assert` like below, although we're moving to using `require` instead of `assert` in library code I think.